internal/controller/nodetemplate_controller.go

```diff
@@ -0,0 +1,390 @@
+			)
+			err = r.Client.Update(ctx, &node)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		// remove finalizer
+		nodeTemplate.Finalizers = nil
+		log.Info("removing finalizer", "NodeTemplate", nodeTemplate.Name, "finalizer", nodeTemplate.Finalizers)
+		err = r.Update(ctx, nodeTemplate)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		// remove metrics
+		r.removeMetrics(nodeTemplate)
+		return ctrl.Result{}, nil
+	}
+
+	defer func() {
+		updateStatusErr := r.updateStatus(ctx, nodeTemplate)
+		if updateStatusErr != nil {
+			err = updateStatusErr
+		}
+		// if nodeTemplate.Status.Sufficient is not true, requeue after 1 second
+		if !nodeTemplate.Status.Sufficient {
+			result = ctrl.Result{
+				RequeueAfter: time.Second * 1,
+			}
+		}
+	}()
+
+	// sync existing nodes
+	for _, node := range existingNodes.Items {
+		nodeOriginal := node.DeepCopy()
+		r.reflectNodeConfiguration(&node, nodeTemplate)
+		if !reflect.DeepEqual(*nodeOriginal, node) {
+			log.Info("updating node",
+				"node", node.Name,
+				"NodeTemplate", nodeTemplate.Name,
+			)
+			if nodeTemplate.Spec.DryRun {
+				log.Info("dry run mode, skipping update nodes.\n diff:\n"+calcDiff(nodeOriginal, &node),
+					"node", node.Name,
+					"NodeTemplate", nodeTemplate.Name,
+					"phase", "updating",
+				)
+				continue
+			}
+			err = r.Client.Update(ctx, &node)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	// if there are NodeTemplates that is lack of nodes with higher priority, skip ensuring of nodes.
+	nodeTemplates := nyallocatorv1.NodeTemplateList{}
+	err = r.Client.List(ctx, &nodeTemplates)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	for _, nt := range nodeTemplates.Items {
+		if nt.Spec.Priority > nodeTemplate.Spec.Priority && !nt.Status.Sufficient {
+			log.Info("skip ensuring node, because there is a NodeTemplate with higher priority that is not sufficient",
+				"target", nodeTemplate.Name,
+				"prior", nt.Name,
+				"current", nt.Status.CurrentNodes,
+				"expected", nt.Spec.Nodes,
+			)
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// allocate new nodes
+	allNode := corev1.NodeList{}
+	spareNodes := map[string]corev1.Node{}
+	err = r.Client.List(ctx, &allNode)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, node := range allNode.Items {
+		for _, taint := range node.Spec.Taints {
+			if taint.Key == SpareTaintKey {
+				selector, err := metav1.LabelSelectorAsSelector(nodeTemplate.Spec.Selector)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+				if selector.Matches(labels.Set(node.Labels)) {
+					spareNodes[node.Name] = node
+				}
+			}
+		}
+	}
+
+	countNodeByZone := make(map[string]int)
+	for _, n := range existingNodes.Items {
+		if zone, ok := n.Labels[ZoneLabelKey]; ok {
+			countNodeByZone[zone]++
+		}
+	}
+
+	for i := len(existingNodes.Items); i < nodeTemplate.Spec.Nodes; i++ {
+		if len(spareNodes) == 0 {
+			log.Info("no spare nodes found", "target", nodeTemplate.Name)
```

Comment:
In other logs, `NodeTemplate` is used.
```suggestion
			log.Info("no spare nodes found", "NodeTemplate", nodeTemplate.Name)
```